### PR TITLE
Add cache for GetTypeByMetadataName

### PR DESCRIPTION
**Pull request by agocke**
This seems to be a popular call among analyzers, especially XUnit. It's a
reasonable way to acquire a well-known analyzer type, but analyzers often
don't save the resulting symbol. This causes significant CPU work to be done
repeatedly as XUnit looks for the same symbol over and over, passing in the
same type name each time.

A very small, simple cache seems to solve the problem without requiring any
changes on the XUnit side.

### CODE REVIEW

File: src/Compilers/Core/Portable/Collections/ConcurrentCache.cs [hunk 1]
```diff
@@ -16,6 +16,8 @@ namespace Microsoft.CodeAnalysis
     internal class ConcurrentCache<TKey, TValue> :
         CachingBase<ConcurrentCache<TKey, TValue>.Entry> where TKey : IEquatable<TKey>
     {
+        private readonly IEqualityComparer<TKey> _keyComparer;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/ConcurrentCache.cs]: why was this needed?  it doesn't seem strictly necessary as part of this PR.

Comment (agocke) [reply src/Compilers/Core/Portable/Collections/ConcurrentCache.cs]: Need to pass a key comparer

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 2]
```diff
@@ -902,6 +902,15 @@ public IPointerTypeSymbol CreatePointerTypeSymbol(ITypeSymbol pointedAtType)
...
         protected abstract IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol elementType);
 
+        // PERF: Traces show that analyzers may use this method frequently, often requesting
+        // the same symbol over and over again. In certain instances this can even consume 1% of
+        // the total CPU time of compilation. This is an extremely simple cache that evicts on
+        // hash code conflicts, but seems to do the trick. The size is mostly arbitrary. My guess
+        // is that there are maybe a couple dozen analyzers in the solution and each one has
+        // ~0-2 unique well-known types, and the chance of hash collision is very low.
+        private ConcurrentCache<string, INamedTypeSymbol> _getTypeCache =
+            new ConcurrentCache<string, INamedTypeSymbol>(50);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: but you don't actually pass a comparer here.  which means you just get `EqualityComparer<TKey>.Default`, which should have the same behavior as the old code (except for null handling).

Comment (agocke) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs]: Woops! Good news: it's already fast enough to drop it off the perf profile. Better news: I can make it even faster!

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs]: Note: i think doing normal .Equals here is good.  I'm not sure people will have the exact same reference for the full name each time they call this.  Having this have value-semantics for the name seems sensible to me :)

Comment (agocke) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs]: Well, one thing I noticed is that a significant amount of time was spent in just calculating the hash code of the string. I'll check and see if it's still hot in the trace (even if the total time in GetTypeByMetadataName is now minimal).

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs]: Works for me :)

Comment (agocke) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs]: OK, when I use default equality, I get 2 samples (out of 67,000) in the hash function. When I move to reference equality, there are now zero samples in the hash. I think I'm going to go with reference equality. I don't really think that value equality is buying us much, because an analyzer allocating a new string for the type name every invocation seems like it has more significant problems.

**Comment by jaredpar** [comment]
Can we add a bit more detail in the commit as to how you determined this was a problem? That way a few years from now a `git blame` call on this code will give a hint to the future developer on the scenarios where this was an improvement.

File: src/Compilers/Core/Portable/Collections/ConcurrentCache.cs [hunk 3]
```diff
@@ -16,6 +16,8 @@ namespace Microsoft.CodeAnalysis
     internal class ConcurrentCache<TKey, TValue> :
         CachingBase<ConcurrentCache<TKey, TValue>.Entry> where TKey : IEquatable<TKey>
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/ConcurrentCache.cs]: Is this constraint needed now that we take an explicit comparer?

**Comment by agocke** [comment]
@jaredpar Not sure how much more info I should give aside from "traces show this is a CPU hotspot"

**Comment by jaredpar** [comment]
What about the analyzers you were using that demonstrated this was a hot spot?
